### ISSUE TITLE
[HOLD for payment 2023-06-01] [$1000] Web - Chat - In offline mode, on deleting any latest message, if last message has & sign in it, app displays it as &amp; in LHN

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17658]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Action Performed:
2. Open the app
3. Open any report
4. Send message with & sign in it. eg: Hello & welcome
5. Send any other message to that report
6. Switch off connection or go to settings->preferences->force offline
7. Again open that report
8. Delete the last message and observe in LHN

## Expected Result:
App should not convert `&` to `&amp`; in offline mode in LHN

## Actual Result:
App converts `&` to `&amp`; in offline mode in LHN if we delete the message after it

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.1.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/233105111-8835b183-199c-4a5a-863b-278974217c6c.mp4

https://user-images.githubusercontent.com/93399543/233105199-7a5ac1b4-4073-4cd4-b295-3889cd9f0f91.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @dhanashree-sawant

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681808180603719

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01206f24c549758b9a</li>
        <li>Upwork Job ID: 1650576764427841536</li>
        <li>Last Price Increase: 2023-05-15</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @anmurali (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01206f24c549758b9a

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @youssef-lr (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 In offline mode, on deleting any latest message, if last message has & sign in it, app displays it as &amp; in LHN
### What is the root cause of that problem?
This is caused because there is a bug in `ExpensiMark`, when we are parsing text to html we are escaping it: 
https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L331-L333

But when we are doing reverse html to string we are not unescaping it:
https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L540-L548

This is also visible when we add new comment. In optimistic data, we are generating this:
<img width="540" alt="Screenshot 2023-04-25 at 1 49 15 AM" src="https://user-images.githubusercontent.com/59907218/234113623-bea8869c-d547-4f41-b932-ab7fd8e70dd4.png">
And backend is pushing this:
<img width="487" alt="Screenshot 2023-04-25 at 1 50 12 AM" src="https://user-images.githubusercontent.com/59907218/234113814-23ce3e40-a4d2-4f96-8b06-b04c271535f8.png">

When we delete and generate optimistic data, we are using `htmlToText` to generate lastMessageText:
https://github.com/Expensify/App/blob/dd827e953adb07495c42256a36017c56f30348d2/src/libs/ReportActionsUtils.js#L160
and since it is offline it is keeping optimistic data and showing its value. 
### What changes do you think we should make in order to solve the problem?
We should update `ExpsiMark.htmlToText` to unescape using `_.unescape` here:
https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L540-L542

### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@alitoshmatov Thanks for the proposal. The RCA is correct. The fix is almost correct. We should call `_.unescape` after applying the `htmlToTextRules` and not before, otherwise html will get corrupted. I won't block on that though.

:ribbon: :eyes: :ribbon: C+ reviewed

cc @youssef-lr

**Comment by youssef-lr** [comment]
Proposal looks to be working. However, I was curious about what we do differently when setting the report's `lastMessageText` optimistically that already unescapes html, and I found out that we are using `Str.htmlDecode` for this [here](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js#L209). I think we should keep things consistent and reuse the same method. So instead of modifying `htmlToText` we can use `Str.htmlDecode` in `getLastVisibleMessageText`.

What do you think @s77rt?

**Comment by s77rt** [comment]
@youssef-lr There have been a number of bugs related to html <-> text encoding and by the looks of it we are not exactly done with those :sweat_smile: so I'm not sure why we are using both `Str.htmlDecode` and `parser.htmlToText`. 

___

> So instead of modifying `htmlToText` we can use `Str.htmlDecode` in `getLastVisibleMessageText`.

I think we should fix still fix `parser.htmlToText` even if we are not going to use it here.

___

> Also, for this solution to work, we also need to update this [line](https://github.com/Expensify/App/blob/main/src/libs/ReportActionsUtils.js#L158) to use message[0].text instead of message[0].html.

I don't think this is necessary. I think you based this assumption on this [line](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/actions/Report.js#L205) but this piece of [code](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/actions/Report.js#L205-L212) is actually misleading. We should either: 1. Get HTML -> Decode it - OR - 2. Get TEXT (and use it directly). So in that case we should not have bothered calling `Str.htmlDecode` since `lastCommentText` is already a text not html.

___

That point ^ actually raises another concern, in `getLastVisibleMessageText` why we are fetching the html message and decode it? Why not just use the text message itself. It looks we are decoding a message that we already have a decoded version of.

___

Sorry for the lengthy comment. Here is the plan:

1. Ask on Slack if someone knows why we have two html decoder methods
2. Agree on using only one and replace all occurrences of the other - OR - "Unify" them
3. In [getLastVisibleMessageText](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L158) use `text` instead of `html` (and remove the [decoding part](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L159-L160)). 

What do you think?

cc @pecanoro as you worked on similar bugs before.

**Comment by s77rt** [comment]
@youssef-lr ^

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by s77rt** [comment]
I think `Str.htmlDecode` is just another version of `_.unescape`. We are probably using it as a workaround due the bug on `parser.htmlToText` (since it's missing the `_.unescape` call).

Those two functions are different. To convert from html to text we actually need to use `parser.htmlToText`.

So I think the new plan is:
1. Fix the bug on `parser.htmlToText`
2. In [getLastVisibleMessageText](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L158) use `text` instead of `html` (and remove the [decoding part](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L159-L160)).


What do you think? @youssef-lr @pecanoro

**Comment by alitoshmatov** [comment]
> I think Str.htmlDecode is just another version of _.unescape. We are probably using it as a workaround due the bug on parser.htmlToText (since it's missing the _.unescape call).

@s77rt I agree with you

There are multiple instances of using `Str.htmlDecode` with `parser.htmlToText`
https://github.com/Expensify/App/blob/f8daa3f87d09edc0d9492f9c9fc1861276acca88/src/components/CopySelectionHelper.js#L40
https://github.com/Expensify/App/blob/2e8396f72c51a49a74033fadd3dec65cae56be7a/src/pages/home/report/ContextMenu/ContextMenuActions.js#L162
https://github.com/Expensify/App/blob/237c8cde7cc43287d3526fa77ab27d2c6cfd24b8/src/libs/ReportUtils.js#L1125
https://github.com/Expensify/App/blob/b2c28acea914a54dc6585485529fd01019a8bfcf/src/libs/actions/Report.js#L939-L943

**Comment by MelvinBot** [comment]
@youssef-lr @anmurali @s77rt this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by s77rt** [comment]
@MelvinBot We have a fix for the issue already and a plan as well. Just waiting for some :eyes: on this and we should be good to move forward :rocket:!

**Comment by pecanoro** [comment]
@s77rt Good plan, thank you for summarizing!

**Comment by s77rt** [comment]
@youssef-lr I think this is waiting for your input here

**Comment by youssef-lr** [comment]
Sorry I was focused on some high priority tasks that have a tight deadline. I will get to this today. Thanks for the bump @s77rt.

**Comment by s77rt** [comment]
Not overdue. Waiting for @youssef-lr

**Comment by pecanoro** [comment]
@youssef-lr @s77rt Can you retest this? I recently change something related to this on the back-end but I am not sure if it is solved. If we are encoding/decoding in the front-end, the issue might have persisted. Either way, we are sending text and not HTML for the LHN now, so it's something to be taken into account.

**Comment by pecanoro** [comment]
Thanks for clarifying! Just making sure we were not pointlessly waiting 😄

**Comment by youssef-lr** [comment]
>So I think the new plan is:
>1. Fix the bug on `parser.htmlToText`
>2. In [getLastVisibleMessageText](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L158) use `text` instead of `html` (and remove the [decoding part](https://github.com/Expensify/App/blob/1bc672f8f9e7f057a03d041fb1a3b5cdde789c5e/src/libs/ReportActionsUtils.js#L159-L160)).

This sounds good to me. Let's move forward with this @s77rt @alitoshmatov?

**Comment by s77rt** [comment]
@youssef-lr Let's get this fixed :rocket:! Please assign @alitoshmatov.

**Comment by s77rt** [comment]
@youssef-lr ^

**Comment by youssef-lr** [comment]
Not overdue, @alitoshmatov we can move forward with your proposal.

**Comment by alitoshmatov** [comment]
Applied to the job PR will be ready in couple in a day

**Comment by alitoshmatov** [comment]
Opened pr in `expensify-common` - https://github.com/Expensify/expensify-common/pull/538

Will open app pr once it gets merged

**Comment by alitoshmatov** [comment]
PR is ready - https://github.com/Expensify/App/pull/19237

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/expensify-common/pull/440
- The offending PR has been commented on: https://github.com/Expensify/expensify-common/pull/440#issuecomment-1562608507
- .A discussion in #expensify-bugs has been started: https://expensify.slack.com/archives/C049HHMV9SM/p1685008306988269
- Determine if we should create a regression test for this bug: No, we have automatic tests for this one

**Comment by s77rt** [comment]
Sure! I have checked that issue, as far as I understood it, we were already decoding a plain text which is wrong but used to work because our html to text was not functional well. After this PR the html to text is working correctly but now if we decode a plain text it won't work correctly (and it should not).

If this is the case, then yes it's a regression from this one, we should have removed the outdated decode-plaintext-thing. Let us know once this is confirmed @parasharrajat.

**Comment by therealsujitk** [comment]
@s77rt @parasharrajat you can use the PR I created - https://github.com/Expensify/expensify-common/pull/543, I don't want it to go to waste. I don't expect to be compensated since I wasn't assigned to it.

**Comment by s77rt** [comment]
@alitoshmatov Can you please draft a follow up PR to remove the redundant decode call here https://github.com/Expensify/App/blob/351056286bbab7372831a4522651eacfea4c5cc9/src/libs/ReportUtils.js#L667

**Comment by therealsujitk** [comment]
@s77rt there are multiple places where this needs to be done, just letting you know.

**Comment by therealsujitk** [comment]
@parasharrajat they are out of scope for #19789 however they are regressions from #19237. You just have to look for all occurrences of `Str.htmlDecode()` and remove them if `parser.htmlToText()` has been used.

**Comment by alitoshmatov** [comment]
https://github.com/Expensify/App/issues/19789 was present even before my PR(https://github.com/Expensify/App/pull/19237). No matter if we added `_unescape` to `parser.htmlToText` the problem in https://github.com/Expensify/App/issues/19789 would still remain. We can even completely remove `parser.htmlToText` and the problem remains. The problem is caused by `Str.htmlDecode` decoding html codes like `&#21;` which is expected behavior. 
The main question here is whether we should have removed `Str.htmlDecode` completely when we fixed `parser.htmlToText`. At the time we acknowledged that we had [both](https://github.com/Expensify/App/issues/17658#issuecomment-1533282604) `parser.htmlToText` and `Str.htmlDecode`, and we were not entirely sure why we had both of them and assumed they were used for [different purposes](https://github.com/Expensify/App/issues/17658#issuecomment-1533095827).

**Comment by alitoshmatov** [comment]
Even removing `Str.htmlDecode` would have affected a lot of places. I think https://github.com/Expensify/App/pull/19237 achieved its purpose of fixing the current issue, without introducing new bugs.

**Comment by therealsujitk** [comment]
> #19789 was present even before my PR(#19237).

Are you sure, because I tested it and confirmed. Previously `parser.htmlToText()` did not unescape the result before returning so when `text='&lt; &amp;lt;'` `parser.htmlToText(text)` returned `&lt; &amp;lt;` and `Str.htmlDecode(parser.htmlToText(text))` returned `< &lt;`. Now `parser.htmlToText(text)` unescapes the result so it returns `< &lt;` and `Str.htmlDecode(parser.htmlToText(text))` returns `< <` which is incorrect.

**Comment by alitoshmatov** [comment]
@therealsujitk 
I tested with `&#21;` since it was in https://github.com/Expensify/App/issues/19789 description

**Comment by therealsujitk** [comment]
> I tested with `&#21;` since it was in #19789 description

Even if you test it with that, it first get's escaped in the `replace()` function and becomes `&amp;#21;`. Previously it was being decoded only once so it became `&#21;` again and was printed as is. Now it gets decoded twice so `&amp;#21;` -> `&#21;` -> .

**Comment by therealsujitk** [comment]
You can test it by commenting the `_.unescape()` in **ExpensiMark.js**, you'll see this issue wasn't there before.

**Comment by youssef-lr** [comment]
@alitoshmatov @s77rt do we need a follow up PR to fix the regression?

**Comment by s77rt** [comment]
We have this https://github.com/Expensify/App/pull/20443.

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by s77rt** [comment]
Not overdue. This is on hold for payment (Jun 23rd same as https://github.com/Expensify/App/issues/19789)

**Comment by alitoshmatov** [comment]
https://github.com/Expensify/App/issues/19789 is closed, and this one is ready to proceed.

**Comment by alitoshmatov** [comment]
Another friendly bump

**Comment by s77rt** [comment]
Payments:
- Bug reporter @dhairyasenjaliya $250
- Contributor @alitoshmatov $1000 (no bonus due to regression)
- C+ @s77rt $500 (no bonus and -50% due to regression)

**Comment by dhanashree-sawant** [comment]
Hi @s77rt, I think you meant @dhanashree-sawant 😅

**Comment by s77rt** [comment]
Oops my bad. Sorry @dhairyasenjaliya for the false positive :sweat_smile:

**Comment by dhanashree-sawant** [comment]
Hi @anmurali, Thanks! Offer accepted.

**Comment by alitoshmatov** [comment]
Accepted

**Comment by dhanashree-sawant** [comment]
Hi @pecanoro , @anmurali I have accepted offer on upwork but it is still left to be approved, can you check once you are available?

**Comment by alitoshmatov** [comment]
Same here

**Comment by pecanoro** [comment]
@anmurali Can you approve the offers, please? 🤗
